eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/AssignedOnceVariableToBeFinalCheck.java

```diff
@@ -0,0 +1,267 @@
+            final String fieldName = field.findFirstToken(TokenTypes.IDENT).getText();
+            if (assignmentsFromConstructor.contains(fieldName) && !assignmentsFromMethods.contains(fieldName)) {
+                log(field, String.format(ERROR_MSG, fieldName));
+            }
+        }
+    }
+
+    /*
+    * Get the field token from an assignation token.
+    * This method handles cases for fields referenced as `this.field` or only `field`
+    * It will get parameters from the method definition to ignore assignations to those parameters
+    */
+    private DetailAST getAssignedField(final DetailAST assignationToken) {
+        final Set<String> scopeParentParameterSet = getParameterSet(scopeParent.findFirstToken(
+            TokenTypes.PARAMETERS));
+        final DetailAST firstChild = assignationToken.getFirstChild();
+        final DetailAST assignationWithDot = firstChild.getType() == TokenTypes.DOT ? firstChild : null;
+
+        if (assignationWithDot != null) {
+            if (assignationWithDot.branchContains(TokenTypes.LITERAL_THIS)) {
+                return assignationWithDot.findFirstToken(TokenTypes.IDENT);
+            }
+        } else {
+            final DetailAST variableNameToken = assignationToken.getFirstChild();
+            // make sure the assignation is not for a method parameter
+            if (!scopeParentParameterSet.contains(variableNameToken.getText())) {
+                return variableNameToken;
+            }
+        }
+
+        return null;
+    }
+
+    /*
+    * Saves a field name to a container depending on the provided type
+    */
+    private void saveField(String fieldName, int scopeParentType) {
+        if (scopeParentType == TokenTypes.METHOD_DEF) {
+            assignmentsFromMethods.add(fieldName);
+        } else if (scopeParentType == TokenTypes.CTOR_DEF) {
+            assignmentsFromConstructor.add(fieldName);
+        }
+    }
+
+    /*
+     * Review an assignation to save fields that gets assigned in constructor or in any method
+     *
+     * @param assignationToken an assignation token
+     */
+    private void checkAssignation(final DetailAST assignationToken) {
+        if (scopeParent == null || assignationToken.getChildCount() == 0) {
+            // not inside any method or constructor definition. No need to check anything
+            // or this is an assignation from a notation like @Test(timeout = 5000) where assignation has not ChildCount
+            return;
+        }
+
+        final DetailAST assignationParent = assignationToken.getParent();
+        if (assignationParent != null && TokenTypes.VARIABLE_DEF == assignationParent.getType()) {
+            // Assignation for a variable definition. No need to check this assignation
+            return;
+        }
+
+        DetailAST fieldToken = getAssignedField(assignationToken);
+
+        if (fieldToken != null) {
+            saveField(fieldToken.getText(), scopeParent.getType());
```

Comment:
Does this check work in the following cases?

1. Variable is assigned in class initialization not constructor but no where else.
    ```java
    class Foo {
        private String myBar = "This is a bar";

        Foo(String something) {
            //... other things
        }
    }
    ```
1. Does this work in the case where a method contains a variable with the same name as a member variable?
    ```java
    class Foo {
        private String myBar = "This is a bar";

        private void bar() {
            String myBar = "something";
            myBar = "something else";
            System.out.println(myBar);
        }
    }
    ```

Reply:
for case 1, check won't ask field to be final. Think that was not asked for the rule.
It wouldn't be too much trouble to have it added... would you like to have it?  I could make it configurable maybe, with a settingArgument...

for case 2, current check only validate the in-parameters for the method.  I need to fix it to also get any local variable declaration. Very good point!  It didn't find any case like this, but it won't be hard to add. thanks

Reply:
Actually, for case 2. I would rather add a new checkstyle rule to require people to never use local variables with the same name as a field. IMO, this is not a good practice and make code complicated to read/understand.

imagine next code
```
public class RecordedData {
    private type variable;
public method(a parameter) {
 variable = null;
        try{
            type variables = null;
            variables = null;
        } catch (Exception e) {
            ..
        }
        variables = null;
```
This is valid, but our current rule would need to handle temporal scopes,, to realize when we are inside any block that creates a new scope like a try/catch/Closeable , etc..  XD.

Let me know whay you think @conniey . We don't currently have anything like this in our code though

Reply:
For case 1..
if field is only assigned in class initialization, shouldn't we ask for `static final` ?  Since it looks like it won't have a per-instance value anyway.. so, I believe better to ask for static final on those cases

Reply:
1. No, shouldn't be `static final`. Instantiations like the one below are valid.
    ```java
    class EventHubConsumer {
        private AtomicBoolean isDisposed = new AtomicBoolean();
    }
    ```
1. One option is to keep track of variable creations that also have the same name as any member variables. If it does, and there is an assignment that uses `this.`, then it is not a final variable. Otherwise, if there are no variable creations with the same name, and it does an assignment, then it is used.

Regarding your new rule: Well, setter methods like the one below are valid. Your new rule where local variables can't share the same local variable name may flag it.
    ```java
    private Foo foo;
    private void setSomething(Foo foo) {
        this.foo = foo;
    }
    ```

Reply:
@conniey , number 1 has been implemented.  We now check if a field is initialized and its value is never changed, we ask it to be final.